src/hotspot/share/classfile/protectionDomainCache.hpp

```diff
@@ -28,88 +28,42 @@
 #include "oops/oop.hpp"
 #include "oops/weakHandle.hpp"
 #include "runtime/atomic.hpp"
-#include "utilities/hashtable.hpp"
 
-// This class caches the approved protection domains that can access loaded classes.
-// Dictionary entry pd_set point to entries in this hashtable.   Please refer
-// to dictionary.hpp pd_set for more information about how protection domain entries
-// are used.
-// This table is walked during GC, rather than the class loader data graph dictionaries.
-class ProtectionDomainCacheEntry : public HashtableEntry<WeakHandle, mtClass> {
-  friend class VMStructs;
- public:
-  oop object();
-  oop object_no_keepalive();
-
-  ProtectionDomainCacheEntry* next() {
-    return (ProtectionDomainCacheEntry*)HashtableEntry<WeakHandle, mtClass>::next();
-  }
-
-  ProtectionDomainCacheEntry** next_addr() {
-    return (ProtectionDomainCacheEntry**)HashtableEntry<WeakHandle, mtClass>::next_addr();
-  }
-
-  void verify();
-};
-
-// The ProtectionDomainCacheTable contains all protection domain oops. The
-// dictionary entries reference its entries instead of having references to oops
-// directly.
-// This is used to speed up system dictionary iteration: the oops in the
-// protection domain are the only ones referring the Java heap. So when there is
-// need to update these, instead of going over every entry of the system dictionary,
-// we only need to iterate over this set.
+// The ProtectionDomainCacheTable contains all protection domain oops as WeakHandles.
 // The amount of different protection domains used is typically magnitudes smaller
 // than the number of system dictionary entries (loaded classes).
-class ProtectionDomainCacheTable : public Hashtable<WeakHandle, mtClass> {
-  ProtectionDomainCacheEntry* bucket(int i) const {
-    return (ProtectionDomainCacheEntry*) Hashtable<WeakHandle, mtClass>::bucket(i);
-  }
-
-  // The following method is not MT-safe and must be done under lock.
-  ProtectionDomainCacheEntry** bucket_addr(int i) {
-    return (ProtectionDomainCacheEntry**) Hashtable<WeakHandle, mtClass>::bucket_addr(i);
-  }
-
-  ProtectionDomainCacheEntry* new_entry(unsigned int hash, WeakHandle protection_domain) {
-    ProtectionDomainCacheEntry* entry = (ProtectionDomainCacheEntry*)
-      Hashtable<WeakHandle, mtClass>::new_entry(hash, protection_domain);
-    return entry;
-  }
-
-  static unsigned int compute_hash(Handle protection_domain);
-
-  int index_for(Handle protection_domain);
-  ProtectionDomainCacheEntry* add_entry(int index, unsigned int hash, Handle protection_domain);
-  ProtectionDomainCacheEntry* find_entry(int index, Handle protection_domain);
+class ProtectionDomainCacheTable : public AllStatic {
```

Comment:
How about adding a comment to say what this table maps from/to? Something like:
```
// The ProtectionDomainCacheTable maps all java.security.ProtectionDomain objects that are
// registered by DictionaryEntry::add_protection_domain() to a unique WeakHandle. 
```

Reply:
Okay, I added your comment.

Reply:
I still can't decipher what this "map" actually holds. What does the unique WeakHandle actually refer to?

Reply:
I've suggested a different comment above.